app/src/androidTest/java/br/com/simplepass/loadingbutton/ViewVisibilityIdlingResource.kt

```diff
@@ -0,0 +1,40 @@
+package br.com.simplepass.loadingbutton
+
+import android.view.View
+import androidx.test.espresso.IdlingResource
+import androidx.test.espresso.IdlingResource.ResourceCallback
+
+/*
+    IdlingResource that waits until the specified view matches the expected visibility criteria
+    Converted to Kotlin from original implementation found here: https://stackoverflow.com/a/35080198
+ */
+class ViewVisibilityIdlingResource(private val view: View, private val expectedVisibility: Int) : IdlingResource {
```

Comment:
I just saw someone copy this over into another project recently. In my experience, using IdlingResource for views can be hokey and is explicitly against the Espresso guidelines (see: https://developer.android.com/training/testing/espresso/idling-resource#create-own)

My preference is towards a solution like this:

```
package com.homeaway.android.test

import android.view.View
import androidx.test.espresso.Espresso
import androidx.test.espresso.FailureHandler
import androidx.test.espresso.Root
import androidx.test.espresso.ViewAssertion
import androidx.test.espresso.matcher.RootMatchers
import org.hamcrest.Matcher
import org.joda.time.DateTimeUtils

class Wait private constructor(
        private val delay: Long,
        private val duration: Long,
        private val startTime: Long,
        private val viewMatcher: Matcher<View>,
        private val rootMatcher: Matcher<Root>,
        private val assertion: ViewAssertion): FailureHandler {

    init {
        checkAssertion()
    }

    override fun handle(error: Throwable?, viewMatcher: Matcher<View>?) {
        val currentTime = DateTimeUtils.currentTimeMillis()

        if (currentTime > (startTime + duration)) {
            throw WaitingTimeoutException(error)
        }

        try {
            Thread.sleep(delay)
        } catch (e: InterruptedException) {
            // let it go
        }

        checkAssertion()
    }

    private fun checkAssertion() {
        Espresso.onView(viewMatcher)
                .inRoot(rootMatcher)
                .withFailureHandler(this)
                .check(assertion);
    }

    class Builder constructor(private val viewMatcher: Matcher<View>) {
        private var delay: Long = 50
        private var duration: Long = 5000
        private var rootMatcher: Matcher<Root> = RootMatchers.DEFAULT

        fun inRoot(matcher: Matcher<Root>): Builder {
            this.rootMatcher = matcher;
            return this
        }

        fun delay(delay: Long): Builder {
            this.delay = delay
            return this
        }

        fun duration(duration: Long): Builder {
            this.duration = duration
            return this
        }

        fun until(assertion: ViewAssertion): Wait {
            val startTime = DateTimeUtils.currentTimeMillis()
            return Wait(delay, duration, startTime, viewMatcher, rootMatcher, assertion)
        }
    }

    companion object {
        @JvmStatic
        fun onView(viewMatcher: Matcher<View>): Builder {
            return Builder(viewMatcher)
        }
    }
}

class WaitingTimeoutException(cause: Throwable?) : RuntimeException(cause)
```

Reply:
This is a loop with assertions in time to time, right? Uhm, for what I understand this will be slower than some version of TextInTextViewIdlyingResource... For me is looks a more natural solution. WDYT?

Reply:
Well, you could set the delay to 1ms which would make a pretty tight loop.
That said, I personally would be less worried about optimal test speed and
instead settle for quick, functional, stable, and reproducible test runs.
I'm not opposed to either approach -- my inclination would be for the one
with minimal dev cost satisfying the above requirements. I have always
tried to avoid doing any view related waiting in an IdlingResource due to
the notes in the documentation and the fact that an IdlingResource should not
block the main thread.

Reply:
The IdlingResources are not blocking the UI thread. 

There are also notes about implementing retry wrappers (in the section `Identify when idling resources are needed`). 

I believe idling resources would a good solution since it is what a developer would expect when opening a PR in this part of the code. I saying this based on the fact that it is what I most commonly see. 

About the note:

> For example, the idling resources that you implement and register shouldn't contain references to View objects.

That makes sense when developing an Android app, since you should not wait for animations, you should test only business logic. But in the case of this library, the only place with any state is the View itself, so I don't think having a View inside the IdlingResource would be a problem to our use case. 

Nevertheless both solution would work, I can accept any of then.

Reply:
Agree, the IdlingResources are not blocking the UI thread here -- what I meant by that comment is that if you block the main thread in an idling resource, you can create deadlock in the test suite. W.r.t. the retry wrapper bit -- I see your point, my opinions has historically been something like "because this Wait uses the Espresso-provided error handling API, it's kosher". This opinion may be worth revisiting. 

I think you've got a valuable opinion here and delegate to you/Carlos on this matter. 

Another thing worth noting -- now that we are talking about this Espresso test, it's worth asking "when do these tests run?". Looking at the travis config, I don't think the androidTest artifacts are currently being compiled are run. Should that be added in the scope of this PR? I would lean towards the answer "yes", and suggest that the travis build should be updated to produce the test apk and run tests on Firebase Test Lab.

Reply:
Well, right now the tests would not run on the CI since there's no configuration for it. 

It would be great if you guys add it to the project. Firebase Test Lab is not an option, because it's not for free and I can't pay for it since I don't make any profit with this project. But it is still possible to run those tests in a emulator running in Travis.

it does make sense for add it in the scope of this PR, but I personally recommend doing it in another one so we don't put more requests in @cavega than he signed for.

Reply:
By the way... there's an example of how to do it here: https://github.com/ebanx/swipe-button

Reply:
How about this:

1. I will update this build in a separate PR to run the tests.
2. Since travis supports emulators, I'll evaluate that first (nice information for myself, I would note).
3. If we had to end up at a paid solution such as FTL (based on the info you've given me, we won't) we could find a way to cover this for you -- for example, we could talk with our company about sponsoring this cost. 

I'll move forward with adding tests to the build once this PR is finalized and merged.

Reply:
Seams great =]. 
If you don't want to wait for this PR to be merged, there's no problem to add a mock integration test just to confirm that it works. But that's up to you. 
Thanks for the help!